src/qbe_ir_generator.cpp

```diff
@@ -644,28 +715,41 @@ void QbeIrGenerator::Visit(const IdExprNode& id_expr) {
     num_recorder.Record(res_num);
     return;
   }
-  assert(id_to_num.count(id_expr.id) != 0);
-  /// @brief Plays the role of a "pointer". Its value has to be loaded to
-  /// the register before use.
-  int id_num = id_to_num.at(id_expr.id);
-  int reg_num = NextLocalNum();
-  if (id_expr.type->IsPtr() || id_expr.type->IsFunc()) {
-    WriteInstr_("{} =l loadl {}", FuncScopeTemp{reg_num},
-                FuncScopeTemp{id_num});
-  } else {
+  if (id_expr.is_global) {
+    int reg_num = NextLocalNum();
     WriteInstr_("{} =w loadw {}", FuncScopeTemp{reg_num},
-                FuncScopeTemp{id_num});
+                user_defined::GlobalPointer{id_expr.id});
+    num_recorder.Record(reg_num);
+    reg_num_to_id[reg_num] = id_expr.id;
+  } else {
+    assert(id_to_num.count(id_expr.id) != 0);
+    /// @brief Plays the role of a "pointer". Its value has to be loaded to
+    /// the register before use.
+    int id_num = id_to_num.at(id_expr.id);
+    int reg_num = NextLocalNum();
+    if (id_expr.type->IsPtr() || id_expr.type->IsFunc()) {
+      WriteInstr_("{} =l loadl {}", FuncScopeTemp{reg_num},
+                  FuncScopeTemp{id_num});
+    } else {
+      WriteInstr_("{} =w loadw {}", FuncScopeTemp{reg_num},
+                  FuncScopeTemp{id_num});
+    }
+    num_recorder.Record(reg_num);
+    // Map the temporary reg_num to id_num, so that upper level nodes can store
+    // value to id_num instead of reg_num.
+    reg_num_to_id_num[reg_num] = id_num;
   }
-  num_recorder.Record(reg_num);
-  // Map the temporary reg_num to id_num, so that upper level nodes can store
-  // value to id_num instead of reg_num.
-  reg_num_to_id_num[reg_num] = id_num;
 }
 
 void QbeIrGenerator::Visit(const IntConstExprNode& int_expr) {
-  int num = NextLocalNum();
-  WriteInstr_("{} =w copy {}", FuncScopeTemp{num}, int_expr.val);
-  num_recorder.Record(num);
+  if (int_expr.is_global) {
+    global_var_init_vals.push_back(
+        {int_expr.val, std::make_shared<PrimType>(PrimitiveType::kInt)});
```

Comment:
```suggestion
    global_var_init_vals.emplace_back(
        int_expr.val, std::make_shared<PrimType>(PrimitiveType::kInt));
```

Reply:
It'll throw this error after changing to emplace back, I'll keep the original implementation.
```
/usr/include/c++/12/bits/new_allocator.h:175:11: error: no matching function for call to ‘{anonymous}::GlobalVarInitVal::GlobalVarInitVal(const int&, std::shared_ptr<PrimType>)’
  175 |         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Oh, this is because `GlobalVarInitVal` only has compiler-provided constructors, which are the default constructor and the copy/move constructors. Never mind, keeping the original one is totally fine. 👍